src/PairedUrlStructure/LegacyReaderUrlStructure.php

```diff
@@ -25,8 +25,10 @@ final class LegacyReaderUrlStructure extends PairedUrlStructure {
 	 * @return string AMP URL.
 	 */
 	public function add_endpoint( $url ) {
-		$post_id = $this->url_to_postid( $url );
+		// Make sure any existing AMP endpoint is removed.
+		$url = $this->remove_endpoint( $url );
 
+		$post_id = $this->url_to_postid( $url );
```

Comment:
Previously when an AMP endpoint was being supplied, `$post_id` would always be `0` since it wouldn't be recognized. So removing the endpoint first ensures that `url_to_postid` works as expected.

Reply:
I've added the missing assertions which would have caught this in 28b3fbfc056.